src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs

```diff
@@ -517,10 +575,94 @@ public int MultiDimensionalArrayRank
             }
         }
 
+        public bool IsValueType
+        {
+            get
+            {
+                return (Flags & enum_flag_Category_ValueType_Mask) == enum_flag_Category_ValueType;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public TypeHandle GetArrayElementTypeHandle()
+        {
+            return *(TypeHandle*)&((MethodTable*)Unsafe.AsPointer(ref this))->ElementTypeHandle;
+        }
+
         [MethodImpl(MethodImplOptions.InternalCall)]
         public extern uint GetNumInstanceFieldBytes();
     }
 
+    /// <summary>
+    /// A type handle, which can wrap either a pointer to a <see cref="TypeDesc"/> or to a <see cref="MethodTable"/>.
+    /// </summary>
+    [StructLayout(LayoutKind.Explicit)]
+    internal unsafe struct TypeHandle
+    {
+        // Subset of src\vm\typehandle.h
+
+        /// <summary>
+        /// The address of the current type handle object.
+        /// </summary>
+        [FieldOffset(0)]
+        private readonly void* m_asTAddr;
+
+        /// <summary>
+        /// Gets whether or not this <see cref="TypeHandle"/> wraps a <see cref="TypeDesc"/> pointer.
+        /// If so, <see cref="AsTypeDesc"/> is safe to call. Otherwise, this instance wraps a <see cref="MethodTable"/> pointer.
+        /// </summary>
+        public bool IsTypeDesc
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => ((nint)m_asTAddr & 2) != 0;
+        }
+
+        /// <summary>
+        /// Gets the <see cref="TypeDesc"/> pointer wrapped by the current instance.
+        /// </summary>
+        /// <remarks>This is only safe to call if <see cref="IsTypeDesc"/> returned <see langword="true"/>.</remarks>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public TypeDesc* AsTypeDesc()
+        {
+            Debug.Assert(IsTypeDesc());
+
+            return (TypeDesc*)(m_asTAddr - 2);
+        }
+
+        /// <summary>
+        /// Gets the <see cref="MethodTable"/> pointer wrapped by the current instance.
+        /// </summary>
+        /// <remarks>This is only safe to call if <see cref="IsTypeDesc"/> returned <see langword="false"/>.</remarks>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public MethodTable* AsMethodTable()
+        {
+            Debug.Assert(!IsTypeDesc());
+
+            return (MethodTable*)m_asTAddr;
+        }
+    }
+
+    // Subset of src\vm\typedesc.h
+    [StructLayout(LayoutKind.Explicit)]
+    internal unsafe struct TypeDesc
```

Comment:
`TypeDesc` is not used in the changes. No need to add it.

Reply:
Right. Removed in b68d97678125b9a9c12960ddd5d7dcbddcf1a9a1 🙂